Support random rotations of features

gregor-robinson opened the pull request:
This PR introduces a transformer that rotates input feature space before passing the transformed data to a base learner, with a specific aim of applying random rotations. There is reason to believe that random rotations of feature space can improve the performance of various kinds of learners --- such as by increasing entropy across an ensemble, or by decreasing the incidence of unimportant feature subsets for methods that use ensembles over random feature subsets. One promising application, random rotation ensembles introduced by Blaser & Fryzlewicz, trains each base learner in an ensemble method on a different random rotation of feature space; predictions are performed using the same rotations that were applied to the training data. Blaser & Fryzlewicz (2016) found large improvement in RMSE using this approach on Fisher's iris dataset.

Note: there are a few tests indicated by TODOs that I still want to write before merging this, but I would appreciate eyes on the implementation for feedback. Some questions for the team: What's an appropriate way to handle None in `getGradient`? Are there any other cases you want to see in the unit tests?

maxhutch commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,225 @@
+    */
+  override def getGradient(): Option[Seq[Vector[Double]]] = {
+    var baseGradient = baseResult.getGradient()
+    // TODO: how should this handle None gradients?
```
Then it should just return none (which is the behavior below, I believe).

maxhutch commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,225 @@
+  def getDoubleFeatures(rep: Vector[Any]): IndexedSeq[Int] = {
+    rep.indices.collect( i =>
+      rep(i) match {
+        case x if x.isInstanceOf[Double] => i
```
Does `case _: Double => i` not work because of type erasure?

I'd be a bit surprised if this collect syntax works.  Can you test it when there are some categorical inputs?  You can use [this pattern](https://github.com/CitrineInformatics/lolo/blob/develop/src/test/scala/io/citrine/lolo/learners/RandomForestTest.scala#L18) to generate mixed real/categorical input vectors.

gregor-robinson replied:
This was an experiment that I neglected to roll back, trying to fix IntelliJ's complaint that a `PartialFunction[Int,NotInferredB]` is required. I now suspect that complaint reflects a static analysis bug, but I'd appreciate if you could check this spot for potential issues the static analyzer is correctly flagging.

bfolie replied:
`rep.indices.collect` allows you to collect the elements of `rep.indices` that meet some criterion, but `rep.indices` is just a list of indices (0, 1, 2, 3, ...) and even though that's what you're collecting, you're filtering on something else (the type of `rep(i)`).
I think the simplest way to do this is 
```rep.indices.filter(i => rep(i).isInstanceOf[Double])```
If you want to avoid `isInstanceOf`, then you could do
```
rep.zip(rep.indices).collect {
      case (_: Double, i: Int) => i
    }
```

gregor-robinson replied:
I do prefer that. Thanks!

gregor-robinson commented:
I've finished and pushed the tests I had in mind.

bfolie commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  /**
+   * Apply rotation to a vectors.
```
```suggestion
   * Apply rotation to a vector.
```

bfolie commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,225 @@
+  * @param baseModel model to which to delegate prediction on rotated features
+  * @param rotatedFeatures indices of features to rotate
+  * @param trans matrix to apply to features
+  * @tparam T
```
What is T?

maxhutch commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,221 @@
+                                   baseTrainingResult: TrainingResult,
+                                   rotatedFeatures: IndexedSeq[Int],
+                                   trans: DenseMatrix[Double]
+                                  ) extends TrainingResult {
```
Can you implement `getLoss` or `getPredictedVsActual` for any cases?

maxhutch commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,221 @@
+  * @param rotatedFeatures indices of features to rotate
+  * @param trans matrix to apply to features
+  */
+class RotatedFeatureTrainingResult(
```
Why not a case class here too?

maxhutch commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,221 @@
+  }
+}
+
+class MultiTaskFeatureRotator(baseLearner: MultiTaskLearner) extends MultiTaskLearner {
```
Why not a case class here too?

maxhutch commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -0,0 +1,221 @@
+    * @return list of feature indices that are doubles
+    */
+  def getDoubleFeatures(rep: Vector[Any]): IndexedSeq[Int] = {
+    rep.indices.filter(i => rep(i).isInstanceOf[Double])
```
This is fine since `rep: Vector` has O(1) indexing, but know that you could use `.zipWithIndex.filter{case (i, r) => r.isInstanceOf[Double]}.map(_._1)` to stream it if it were just `rep: Seq`.  I'm never sure a priori what will or won't work with erasure, but you might be able to do `.zipWithIndex.collect{case (i, r: double) => i}` or at least `.zipWithIndex.collect{case (i, r) if r.isInstanceOf[Double] => i}` as well.

gregor-robinson replied:
Thanks, so many options! I'll leave it as is, unless you have strong preference toward another variant.

maxhutch commented on src/test/scala/io/citrine/lolo/transformers/FeatureRotatorTest.scala:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  /**
+   * Regression trees should not depend on rotation
```
Regression trees _should_ depend on the rotation, but they should memorize their inputs regardless (if grown to full depth).  Could you add a check that non-training inputs do change?

maxhutch commented on src/test/scala/io/citrine/lolo/transformers/FeatureRotatorTest.scala:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  /**
+   * Classification trees should not depend on rotation
```
Likewise for regression trees.

maxhutch commented on src/test/scala/io/citrine/lolo/transformers/FeatureRotatorTest.scala:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  /**
+   * Test that multitask rotation has the proper performance and invariants
```
Likewise here as well.

gregor-robinson commented:
Thanks @maxhutch! The invariance for training data was by design, but I agree it's important to test non-training data. To that end, I have added tests that verify that the rotated feature models predict incorrect values when passed rotated data (i.e. the data are rotated twice). This has the advantage of letting me reuse those test data for another new test: verify that the base learner acting on the rotated data reproduce the training labels.

I have implemented pass-through methods and added tests for them, and changed all classes to case classes.

maxhutch commented on src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala:
```diff
@@ -85,6 +85,10 @@ class RotatedFeatureTrainingResult(
+
+  override def getLoss(): Option[Double] = baseTrainingResult.getLoss()
+
+  override def getPredictedVsActual(): Option[Seq[(Vector[Any], Any, Any)]] = baseTrainingResult.getPredictedVsActual()
```
I think you need to un-rotate the feature vectors here.

gregor-robinson replied:
You're right. Fixed.

maxhutch commented on src/test/scala/io/citrine/lolo/transformers/FeatureRotatorTest.scala:
```diff
@@ -80,6 +80,22 @@ class FeatureRotatorTest {
+    */
+  @Test
+  def testPassthroughFunctions(): Unit = {
+    val rotatedTrainingResult = FeatureRotator(GuessTheMeanLearner()).train(data)
```
`GuessTheMeanLearner` is probably too trivial of a case, here, since the predicted labels will be uniform.  I'd recommend a single decision tree (since its still fast).

gregor-robinson replied:
Fair enough. Done.
